packages/ckeditor5-dev-tests/lib/utils/automated-tests/getkarmaconfig.js

```diff
@@ -6,67 +6,98 @@
 /* jshint browser: false, node: true, strict: true */
 'use strict';
 
+const fs = require( 'fs' );
 const path = require( 'path' );
+const glob = require( 'glob' );
+const minimatch = require( 'minimatch' );
+const tmp = require( 'tmp' );
+const karmaLogger = require( 'karma/lib/logger.js' );
 const getWebpackConfigForAutomatedTests = require( './getwebpackconfig' );
 const transformFileOptionToTestGlob = require( '../transformfileoptiontotestglob' );
 
-const reporters = [
+const AVAILABLE_REPORTERS = [
 	'mocha',
 	'dots'
 ];
 
-const coverageDir = path.join( process.cwd(), 'coverage' );
-
 /**
  * @param {Object} options
  * @returns {Object}
  */
 module.exports = function getKarmaConfig( options ) {
+	const basePath = process.cwd();
+	const coverageDir = path.join( basePath, 'coverage' );
+	const log = karmaLogger.create( 'config' );
+
 	if ( !Array.isArray( options.files ) || options.files.length === 0 ) {
-		throw new Error( 'Karma requires files to tests. `options.files` has to be non-empty array.' );
+		return log.error( 'Karma requires files to tests. `options.files` has to be non-empty array.' );
 	}
 
-	if ( !reporters.includes( options.reporter ) ) {
-		throw new Error( `Given Mocha reporter is not supported. Available reporters: ${ reporters.join( ', ' ) }.` );
+	if ( !AVAILABLE_REPORTERS.includes( options.reporter ) ) {
+		return log.error( 'Specified reporter is not supported. Available reporters: %s.', AVAILABLE_REPORTERS.join( ', ' ) );
 	}
 
-	const files = options.files.map( file => transformFileOptionToTestGlob( file ) );
-
-	const preprocessorMap = {};
+	const globPatterns = options.files.map( file => transformFileOptionToTestGlob( file ) );
+	const allFiles = [];
 
-	for ( const file of files ) {
-		preprocessorMap[ file ] = [ 'webpack' ];
+	// We want to create a single entry point for Karma.
+	// Every pattern must be resolved before Karma starts work.
+	for ( const singlePattern of globPatterns ) {
+		const files = glob.sync( singlePattern );
 
-		if ( options.sourceMap ) {
-			preprocessorMap[ file ].push( 'sourcemap' );
+		if ( !files.length ) {
+			log.warn( 'Pattern "%s" does not match any file.', singlePattern );
 		}
+
+		allFiles.push( ...files.filter( file => {
+			// Ignore files which are saved in `manual/` directory. There are manual tests.
+			if ( file.match( /manual/ ) ) {
+				return false;
+			}
+
+			// Ignore `_utils` directory as well because there are saved utils for tests.
+			return !minimatch( file, path.join( '**', 'tests', '**', '_utils', '**', '*.js' ) );
```

Comment:
The second path can be stored before the loop to not recreate it hundred of times.